lib/vector-core/src/event/array.rs

```diff
@@ -142,28 +142,28 @@ pub enum EventArray {
 }
 
 impl EventArray {
-    /// Sets the `OutputId` in the metadata for all the events in this array.
-    pub fn set_output_id(&mut self, output_id: &Arc<ComponentKey>) {
+    /// Sets the ID of the source component in the metadata for all events in this array.
+    pub fn set_source_id(&mut self, source_id: &Arc<ComponentKey>) {
         match self {
             EventArray::Logs(logs) => {
                 for log in logs {
-                    log.metadata_mut().set_source_id(Arc::clone(output_id));
+                    log.metadata_mut().set_source_id(Arc::clone(source_id));
```

Comment:
I wish there was a common idiom to avoid the final clone in a loop like this. Here, it's just an extra atomic increment and decrement, which is a relatively trivial cost, but I recall Brian doing a "clone army" trick elsewhere to avoid a larger cost.

Reply:
Yeah, I think it's probably using something like the pattern I saw in a recent PR of `itertools` and one of the helpers that tells you if the item being iterated was the last in the iterator or not.

I agree, though... suboptimal as it stands.

Reply:
I'm curious to see that itertools PR you reference. Looking at that package, I am not seeing anything that does a smart clone like described.

Reply:
Looks like I wasn't remembering exactly what the change was, but: https://github.com/vectordotdev/vector/blob/master/src/sinks/util/encoding.rs#L40-L63

Basically, was thinking an approach could sort of be based on that logic being shown there.... using the knowledge of where you are in the iteration of the iterator to do some finagling with your value-to-be-cloned-or-consumed.

Reply:
Ah, thank you, that's interesting. I don't know that I had seen `with_position` before.